### PR TITLE
fix(tabs-extended): confusing behavior when in mobile the accordion c…

### PR DESCRIPTION
**Pull request by bruno-amorim**
…annot be closed

### Related Ticket(s)

[ADCMS-4434](https://jsw.ibm.com/browse/ADCMS-4434)

### Description

The design for the Tabs Extended when in mobile is confusing. It turns into an accordion and there's always one item opened and showing. However, that opened item has an arrow on the right hand side, that deceives the user into thinking it's possible to close that item, when in fact, the accordion mimics the desktop behavior for tabs where we always have one item being shown at a time.

**I'm opening this PR initially as draft with the solution of removing the arrow for opened items so the whole team can discuss the best approach here.**

As per discussion with the team, the suggested changes were:

1.     Added the ability to open and close the accordion items;
2.     When all accordion items are closed, in desktop view the first tab is selected as default.

### CODE REVIEW

**Comment by areagan1030** [comment]
@bruno-amorim Agree that this is a subpar user experience on mobile devices/sm+md breakpoints. I'm wondering if following the Carbon component pattern would be an improvement by having a dropdown selection tied to initial tab display. This might help solve the weirdness of the tab not appearing properly interactive — and then needing to scan down the content to figure out where the next interaction point might be.

I also noticed there was a full-width tabbed component made for the Consulting work that appears to be using this dropdown pattern: [see Consulting](https://www.ibm.com/consulting). The affordance is rather bad on this given the type sizes.

One addition point of conversation may be CIBM v2 — it appears that the mobile display here keeps the main tab setup and uses the standard overflow mechanism across breakpoints to allow users to access additional tab nav items.

**Comment by andy-blum** [comment]
@areagan1030 The consulting page tabs were built custom by Wipro for AEM, those aren't C4IBM or C4AEM

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (andy-blum) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:91]: can we remove this?

**Review by andy-blum — CHANGES_REQUESTED** [review]
I'm not sure if it's a regression or not, but I'm seeing the behavior here that John fixed in #11554

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (jkaeser) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:217]: This change is what causes the regression Andy mentioned, @bruno-amorim . We need to programmatically set focus on the active item so the browser scrolls it into view.

**Review by jkaeser — CHANGES_REQUESTED** [review]
Hey, @bruno-amorim! Thanks for your work on this one. It's a really tricky component!

As we discussed, can you try keeping all the `selected` state logic in the `DDSTabsExtended` component rather than adding logic to `DDSTab` elements? I was thinking you might be able to achieve this by adding some logic to the `_handleAccordionClick()` method. Maybe something like this?

```ts
private _handleAccordionClick(e) {
    const tab = e.target.closest(`${ddsPrefix}-tab`);
    const targetIndex = tab.getIndex() === this._activeTabIndex
        ? -1
        : tab.getIndex();
    this._handleClick(targetIndex, e);
  }
```

Additionally, the auto-scroll behavior isn't working quite right. It appears to scroll to the previous active item when opening an accordion.

https://github.com/carbon-design-system/carbon-for-ibm-dotcom/assets/12532727/ddb2ceaf-663a-4b13-a1dd-0bd9960dadf4

**Comment by bruno-amorim** [comment]
Hi @jkaeser @andy-blum 

All issues pointed out by you are now fixed. @jkaeser I believe the scroll into view issue has been fixed as I've tested it multiple times (even adding more copy to the story tests) and it worked just like the video from your original PR.

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (jkaeser) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:267]: Hrm, can you help me understand why we need this? 

My understanding is that when we cross over the `lg` breakpoint and `this._isLargerOrGreater` gets updated, we end up re-rendering the whole component. As a result, `this._tabItems` gets updated, which causes another update cycle. At that point we have fresh DOM elements, add we only the event listener if we're below the breakpoint.

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (jkaeser) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:117]: I'm also not sure you need this! We're still setting focus in the `_setActiveItem` method.

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (jkaeser) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:309]: These variables have already been declared at the beginning of this method, so I think you can remove this line.

**Review by jkaeser — CHANGES_REQUESTED** [review]
@bruno-amorim Nice! Functionally the behavior is working exactly as we want. Please see my notes for some questions and code cleanup suggestions.

Thank you so much for digging into this one!

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (bruno-amorim) [reply packages/web-components/src/components/tabs-extended/tabs-extended.ts]: Code's been changed as suggested and it's working! Can you please review it now @jkaeser

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (bruno-amorim) [reply packages/web-components/src/components/tabs-extended/tabs-extended.ts]: Code's been remove as suggested and it's working! Can you please review it now @jkaeser

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (bruno-amorim) [reply packages/web-components/src/components/tabs-extended/tabs-extended.ts]: I have removed that line! Can you please review it now @jkaeser

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (bruno-amorim) [reply packages/web-components/src/components/tabs-extended/tabs-extended.ts:249]: Hey @jkaeser,
I needed to add this piece of code because the accordion items were working the first time they're rendered in mobile, but if you switch the viewport to desktop and then go back to mobile, it stops working. I think the changes in the viewport were not properly attaching the events to the accordion items. When I wrote this code in the updated() method, it fixed that bug by ensuring the proper Event Listeners were added and removed based on the viewport. I took a look at the code several times and could not find out why changing the viewport made the accordion items not to work. Do you see something I'm not seeing? Thanks a lot in advance for all your help!

Comment (jkaeser) [reply packages/web-components/src/components/tabs-extended/tabs-extended.ts:249]: Buckle up, this is a lot 😆 

I tracked the event listeners on a `<dds-tab>` element as I crossed back and forth over the breakpoint. What's actually happening is that any time we cross the threshold, a new copy of the `click` event listener is added. None are ever removed.

All of the accumulated event listeners fire off their callbacks in rapid succession any time an accordion item is clicked. When there is only one listener, we get the expected result. When there is a second, the functionality breaks; the first callback fires and changes `this._activeTabIndex`, and then the second fires and sets it back to the clicked tab's index value. Add a third listener, and things work again. It keeps going like this, with even numbers of listeners working and odd numbers failing.

Screenshot 2024-03-15 at 3 01 58 PM

The real fix to all of this is making sure only one copy of the event listener is ever attached. In my opinion, this can be best achieved by removing this new chunk of code and modifying the existing logic. Two potential approaches come to mind:

1. A few lines down in the conditional before adding the event listener, you can check for the absence of an attribute (e.g. `dds-processed`) on the `<dds-tab>` elements, then set that attribute after adding the event listener. This way the conditional will only succeed once.
   - ```ts
      if (!tab.hasAttribute('dds-processed') && !_isLargeOrGreater) {
        tab.addEventListener(...);
        tab.setAttribute('dds-processed', true);
      }
      ```
2. You can remove the existing `tab.addEventListener` and instead use an approach called [Event Delegation](https://javascript.info/event-delegation). You'll attach a single click handler to the `<dds-tabs-extended>` element, check if the element is what we're actually looking for, and only then fire off `this._handleAccordionClick`.
   - ```ts
      @HostListener('click')
      protected _handleClick(event) {
        const { target } = event;
        if (target.matches('button.bx--accordion__heading') {
          this._handleAccordionClick(event);
        }
      }
      ```

Your call on whichever looks better! As always, let me know if you want to go over any of this together.

**Review by jkaeser — CHANGES_REQUESTED** [review]

**Comment by bruno-amorim** [comment]
@jkaeser Thank you so much for your last notes on this, it's ready for your review!

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (jkaeser) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:239]: Nit: use `ddsPrefix` variable.

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (jkaeser) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:241]: Two nits:
1. Use `ddsPrefix` variable.
2. Since we're using this as a boolean attribute, it's a little cleaner to set the attribute to `''`. See https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute. My mistake for recommending the `true` value before!

**Review by jkaeser — APPROVED** [review]
@bruno-amorim Excellent work 🎉 🎉 🎉 

I left some nits for you, but none of them should block this from merging.

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (kennylam) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:239]:
```suggestion
        if (!tab.hasAttribute(`${ddsPrefix}-processed`) && !_isLargeOrGreater) {
```

File: packages/web-components/src/components/tabs-extended/tabs-extended.ts
Comment (kennylam) [inline packages/web-components/src/components/tabs-extended/tabs-extended.ts:241]:
```suggestion
          tab.setAttribute(`${ddsPrefix}-processed`, '');
```

**Review by kennylam — APPROVED** [review]

**Pull request merged**
